File: tests/regenerate/gen_noncentral_distributions.c
/*
 * Reference tables for the noncentral chi-squared and noncentral F
 * distributions (cumchn, cumfnc).
 *
 * Build and run via tests/regenerate/regenerate.sh.
 */

#include "refs/cdflib.h"
#include <stdio.h>
#include <stdlib.h>

static FILE* open_csv(const char* relpath, const char* header) {
    FILE* f = fopen(relpath, "w");
    if (!f) {
        fprintf(stderr, "could not open %s for writing\n", relpath);
        exit(1);
    }
    fprintf(f, "# %s\n", header);
    return f;
}

int main(void) {
    /* ------------------------------------- noncentral chi-squared CDF */
    FILE* f = open_csv("tests/data/chi_squared_noncentral_cdf.csv",
                       "df, ncp, x, cdf, sf");
    double dfs[] = {1.0, 2.0, 5.0, 10.0, 30.0};
    double ncps[] = {0.0, 0.5, 2.0, 10.0, 50.0};
    for (size_t i = 0; i < sizeof(dfs)/sizeof(dfs[0]); i++) {
        for (size_t j = 0; j < sizeof(ncps)/sizeof(ncps[0]); j++) {
            double df = dfs[i], ncp = ncps[j];
            /* Span from below the mean (df + ncp) to several SDs above. */
            double mean = df + ncp;
            double sd = (df + 2.0 * ncp <= 0.0) ? 1.0 :
                        2.0 * (df + 2.0 * ncp);
            sd = (sd <= 0.0) ? 1.0 : sd;
            sd = (sd < 1.0) ? 1.0 : sd;
            double x_max = mean + 5.0 * sd;
            double step = x_max / 30.0;
            if (step <= 0.0) step = 0.1;
            for (double x = step / 4.0; x <= x_max + 1e-12; x += step) {
                double cum = 0.0, ccum = 0.0;
                cumchn(&x, &df, &ncp, &cum, &ccum);
                fprintf(f, "%.17g,%.17g,%.17g,%.17g,%.17g\n",
                        df, ncp, x, cum, ccum);
            }
        }
    }
    fclose(f);

    /* --------------------------------------------- noncentral F CDF */
    f = open_csv("tests/data/fisher_snedecor_noncentral_cdf.csv",
                 "dfn, dfd, ncp, f, cdf, sf");
    double dfns[] = {2.0, 5.0, 10.0, 30.0};
    double dfds[] = {2.0, 5.0, 10.0, 30.0};
    double ncps_f[] = {0.0, 1.0, 5.0, 20.0};
    for (size_t i = 0; i < sizeof(dfns)/sizeof(dfns[0]); i++) {
        for (size_t j = 0; j < sizeof(dfds)/sizeof(dfds[0]); j++) {
            for (size_t k = 0; k < sizeof(ncps_f)/sizeof(ncps_f[0]); k++) {
                double dfn = dfns[i], dfd = dfds[j], ncp = ncps_f[k];
                double f_max = 10.0;
                double step = 0.25;
                for (double fx = step; fx <= f_max + 1e-12; fx += step) {
                    double cum = 0.0, ccum = 0.0;
                    cumfnc(&fx, &dfn, &dfd, &ncp, &cum, &ccum);
                    fprintf(f, "%.17g,%.17g,%.17g,%.17g,%.17g,%.17g\n",
                            dfn, dfd, ncp, fx, cum, ccum);
                }
            }
        }
    }
    fclose(f);

    fprintf(stderr, "wrote 2 tables under tests/data/\n");
    return 0;
}

File: tests/regenerate/gen_noncentral_distributions.f90
! Noncentral chi-squared and noncentral F CDF tables.

program gen_noncentral_distributions
  implicit none
  integer, parameter :: rk = kind(1.0d0)
  external :: cumchn, cumfnc

  real(kind=rk), parameter :: dfs(5)  = (/ 1.0_rk, 2.0_rk, 5.0_rk, 10.0_rk, 30.0_rk /)
  real(kind=rk), parameter :: ncps(8) = (/ &
    0.0_rk, 1.0e-12_rk, 1.0e-10_rk, 1.0e-9_rk, 0.5_rk, 2.0_rk, 10.0_rk, 50.0_rk /)
  real(kind=rk), parameter :: dfns(4) = (/ 2.0_rk, 5.0_rk, 10.0_rk, 30.0_rk /)
  real(kind=rk), parameter :: dfds(4) = (/ 2.0_rk, 5.0_rk, 10.0_rk, 30.0_rk /)
  real(kind=rk), parameter :: ncps_f(7) = (/ &
    0.0_rk, 1.0e-12_rk, 1.0e-10_rk, 1.0e-9_rk, 1.0_rk, 5.0_rk, 20.0_rk /)

  integer :: i, j, k, unit
  real(kind=rk) :: df, ncp, mean, sd, x_max, step, x, cum, ccum
  real(kind=rk) :: dfn, dfd, fx, f_max

  ! noncentral chi-squared CDF
  open(newunit=unit, file='tests/data/chi_squared_noncentral_cdf.csv', &
       status='replace', action='write')
  write(unit, '(a)') '# df, ncp, x, cdf, sf'
  do i = 1, size(dfs)
    do j = 1, size(ncps)
      df = dfs(i)
      ncp = ncps(j)
      mean = df + ncp
      if (df + 2.0_rk * ncp <= 0.0_rk) then
        sd = 1.0_rk
      else
        sd = 2.0_rk * (df + 2.0_rk * ncp)
      end if
      if (sd <= 0.0_rk) sd = 1.0_rk
      if (sd < 1.0_rk)  sd = 1.0_rk
      x_max = mean + 5.0_rk * sd
      step = x_max / 30.0_rk
      if (step <= 0.0_rk) step = 0.1_rk
      x = step / 4.0_rk
      do while (x <= x_max + 1.0e-12_rk)
        call cumchn(x, df, ncp, cum, ccum)
        call putval(unit, df, .false.)
        call putval(unit, ncp, .false.)
        call putval(unit, x, .false.)
        call putval(unit, cum, .false.)
        call putval(unit, ccum, .true.)
        x = x + step
      end do
    end do
  end do
  close(unit)

  ! noncentral F CDF
  open(newunit=unit, file='tests/data/fisher_snedecor_noncentral_cdf.csv', &
       status='replace', action='write')
  write(unit, '(a)') '# dfn, dfd, ncp, f, cdf, sf'
  do i = 1, size(dfns)
    do j = 1, size(dfds)
      do k = 1, size(ncps_f)
        dfn = dfns(i)
        dfd = dfds(j)
        ncp = ncps_f(k)
        f_max = 10.0_rk
        step = 0.25_rk
        fx = step
        do while (fx <= f_max + 1.0e-12_rk)
          call cumfnc(fx, dfn, dfd, ncp, cum, ccum)
          call putval(unit, dfn, .false.)
          call putval(unit, dfd, .false.)
          call putval(unit, ncp, .false.)
          call putval(unit, fx, .false.)
          call putval(unit, cum, .false.)
          call putval(unit, ccum, .true.)
          fx = fx + step
        end do
      end do
    end do
  end do
  close(unit)

  write(0, '(a)') 'wrote 2 tables under tests/data/'

contains
  subroutine putval(unit, v, last)
    integer, intent(in) :: unit
    real(kind=rk), intent(in) :: v
    logical, intent(in) :: last
    character(len=32) :: buf
    write(buf, '(es25.17e3)') v
    if (last) then
      write(unit, '(a)') trim(adjustl(buf))
    else
      write(unit, '(a, a)', advance='no') trim(adjustl(buf)), ','
    end if
  end subroutine putval
end program gen_noncentral_distributions
